### PR TITLE
fix: ping check and model status check

### PR DESCRIPTION
**Pull request by ChuyangDeng**
*Issue #, if available:*

*Description of changes:*
* check model url is valid before sending grpc request to load model
* check model status on server before returning grpc response for load/unload model

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: container/sagemaker/proxy_client.py [hunk 1]
```diff
@@ -64,13 +64,17 @@ def add_model(self, model_name, base_path, model_platform='tensorflow'):
...
+                                                    wait_for_ready=True)
                 self._add_model_to_config_file(model_name, base_path, model_platform)
             except grpc.RpcError as e:
                 if e.code() is grpc.StatusCode.INVALID_ARGUMENT:
                     raise Exception(409, e.details())
+                elif e.code() is grpc.StatusCode.DEADLINE_EXCEEDED:
+                    raise Exception(408, e.details())
                 else:
-                    raise Exception(e.code(), e.details())
+                    raise Exception(500, e.details())
```
Comment (knakad) [inline container/sagemaker/proxy_client.py]: Are we sure we want all elses to throw 500?

File: container/sagemaker/python_service.py [hunk 2]
```diff
@@ -168,21 +172,48 @@ def on_post(self, req, res):
...
+                e = eval(str(e))  # pylint: disable=W0123
+                res.body = e[1].encode('utf-8')
+                if e[0] == 409:
+                    res.status = falcon.HTTP_409
+                    res.body = e[1].encode('utf-8')
+                elif e[0] == 408:
+                    res.status = falcon.HTTP_408
+                    res.body = e[1].encode('utf-8')
+                else:
+                    res.status = falcon.HTTP_500
```
Comment (knakad) [inline container/sagemaker/python_service.py]: I understand that this hasn't changed, but I have the same concern as above. An "else" is not specific enough to determine whether this is actually a 500.

File: container/sagemaker/python_service.py [hunk 3]
```diff
@@ -215,6 +246,60 @@ def _read_model_config(self):
...
+        # model base path doesn't exits
+        if not os.path.exists(model_path):
+            return False
+        # model versions doesn't exist
+        versions = []
+        for _, dirs, _ in os.walk(model_path):
+            for dirname in dirs:
+                if dirname.isdigit():
+                    versions.append(dirname)
+        return bool(versions)
```
Comment (knakad) [inline container/sagemaker/python_service.py]: I'm confused. How does this work? Why are we converting the version to a bool?

File: container/sagemaker/serve.py [hunk 4]
```diff
@@ -314,7 +318,7 @@ def _wait_for_gunicorn(self):
     @contextmanager
     def _timeout(self, seconds):
         def _raise_timeout_error(signum, frame):
-            raise TimeoutError('timed our after {} seconds'.format(seconds))
+            raise Exception('timed our after {} seconds'.format(seconds))
```
Comment (knakad) [inline container/sagemaker/serve.py]: A- Why is this an Exception
B - timed our => timed out *

File: container/sagemaker/proxy_client.py (hunk 1 above)
Comment (ChuyangDeng) [reply container/sagemaker/proxy_client.py]: I'm encoding all other errors to 500 because the TFS server does not return a HTTP error code, instead, it returns the grpc error codes defined here: https://github.com/tensorflow/serving/blob/master/tensorflow_serving/apis/get_model_status.proto

Same as the comment in `python_service.py`

We can however parse each grpc error codes and map them to different http codes.

File: container/sagemaker/python_service.py (hunk 2 above)
Comment (ChuyangDeng) [reply container/sagemaker/python_service.py]: See comment above.

File: container/sagemaker/python_service.py (hunk 3 above)
Comment (ChuyangDeng) [reply container/sagemaker/python_service.py]: This validation method checks if the model file has the correct structure:
1. the model dir exists
2. the model dir contains at least 1 version (aka at least 1 sub folder with version number)

bool(versions) returns False with an empty version list, otherwise it returns True

File: container/sagemaker/serve.py (hunk 4 above)
Comment (ChuyangDeng) [reply container/sagemaker/serve.py]: Updated, raising TimeoutError

File: container/sagemaker/python_service.py (hunk 3 above)
Comment (knakad) [reply container/sagemaker/python_service.py]: Ah, gotcha. Should we make it more obvious what this does by doing something like this?
```
if versions:
    return True
return False
```
_Non-blocking comment._

File: container/sagemaker/proxy_client.py (hunk 1 above)
Comment (knakad) [reply container/sagemaker/proxy_client.py]: Did you link the right file? Those look like status codes.

File: container/sagemaker/serve.py (hunk 4 above)
Comment (knakad) [reply container/sagemaker/serve.py]: [Very Minor] I think we should stick with "Timed out".

Comment (ChuyangDeng) [reply container/sagemaker/serve.py]: Updated.

File: container/sagemaker/python_service.py (hunk 3 above)
Comment (ChuyangDeng) [reply container/sagemaker/python_service.py]: Done :)

File: container/sagemaker/proxy_client.py (hunk 1 above)
Comment (ChuyangDeng) [reply container/sagemaker/proxy_client.py]: You are right. I previously thought falcon response cannot propagate grpc status codes. Looks like it can. Updated.

File: container/sagemaker/tensorflow-serving.js [hunk 5]
```diff
@@ -13,9 +13,68 @@ function invocations(r) {
     }
 }
 
-// ping check does not POST to a specific model, because MME starts with 0 model
+function mme_ping(r) {
```
Comment (jesterhazy) [inline container/sagemaker/tensorflow-serving.js]: this is almost identical to ping_tfs_1_11 - we should refactor to avoid the duplicated code.

File: container/sagemaker/tensorflow-serving.js [hunk 6]
```diff
@@ -13,9 +13,68 @@ function invocations(r) {
...
+    // hack for MME
+    // MME starts without any model loaded and the default-tfs-model is set to "None"
+    var uri = make_tfs_uri(r, true)
+    var options = {
+        method: 'POST',
+        body: '{"instances": "invalid"}'
+    }
+
+    function callback (reply) {
+        if (reply.status == 200 || reply.responseBody.includes('"Servable not found for request: Latest(None)"')) {
```
Comment (jesterhazy) [inline container/sagemaker/tensorflow-serving.js]: can we just check the status code here, or is that not specific enough?

File: container/sagemaker/serve.py [hunk 7]
```diff
@@ -226,13 +227,19 @@ def _setup_gunicorn(self):
     def _create_nginx_config(self):
         template = self._read_nginx_template()
         pattern = re.compile(r'%(\w+)%')
+        ping_request = JS_PING
```
Comment (jesterhazy) [inline container/sagemaker/serve.py]: we could actually check for tfs 1.11 here as well, and assign the ping_tfs_1_11 handler. that would get rid of a request-time if statement in the javascript ping handler.

File: container/sagemaker/python_service.py [hunk 8]
```diff
@@ -168,29 +172,56 @@ def on_post(self, req, res):
...
+                    while True:
+                        model_status = self._get_model_status(model_name)
+                        if self._check_all_versions_available(model_status):
+                            break
+                        sleep(1)
+
+                res.body = msg
+                res.status = falcon.HTTP_200
+            except Exception as e:  # pylint: disable=W0703
+                e = eval(str(e))  # pylint: disable=W0123
```
Comment (jesterhazy) [inline container/sagemaker/python_service.py]: this is dangerous. never use eval on untrusted data. in fact, just never use eval.

File: scripts/start.sh [hunk 9]
```diff
@@ -19,7 +19,6 @@ MODEL_DIR="$(cd "test/resources/models" > /dev/null && pwd)"
 $docker_command run \
     -v "$MODEL_DIR":/opt/ml/model:ro \
     -p 8080:8080 \
-    -e "SAGEMAKER_TFS_DEFAULT_MODEL_NAME=half_plus_three" \
```
Comment (jesterhazy) [inline scripts/start.sh]: not clear what this change has to do with ping

File: docker/1.14/Dockerfile.gpu [hunk 10]
```diff
@@ -2,7 +2,6 @@ FROM nvidia/cuda:10.0-base-ubuntu16.04
 
 LABEL maintainer="Amazon AI"
 LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
-LABEL com.amazonaws.sagemaker.capabilities.multi-models=true
```
Comment (jesterhazy) [inline docker/1.14/Dockerfile.gpu]: 1. not clear what this change has to do with ping
2. does this mean we are removing the mme feature from this version? that's a breaking change.

File: docker/1.14/Dockerfile.eia [hunk 11]
```diff
@@ -1,6 +1,5 @@
 FROM ubuntu:16.04
 LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
-LABEL com.amazonaws.sagemaker.capabilities.multi-models=true
```
Comment (jesterhazy) [inline docker/1.14/Dockerfile.eia]: 1. not clear what this change has to do with ping
2. does this mean we are removing the mme feature from this version? that's a breaking change.

File: docker/1.13/Dockerfile.gpu [hunk 12]
```diff
@@ -2,10 +2,9 @@ FROM nvidia/cuda:10.0-base-ubuntu16.04
 
 LABEL maintainer="Amazon AI"
 LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
-LABEL com.amazonaws.sagemaker.capabilities.multi-models=true
```
Comment (jesterhazy) [inline docker/1.13/Dockerfile.gpu]: 1. not clear what this change has to do with ping
2. does this mean we are removing the mme feature from this version? that's a breaking change.

File: docker/1.13/Dockerfile.eia [hunk 13]
```diff
@@ -1,6 +1,5 @@
 FROM ubuntu:16.04
 LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
-LABEL com.amazonaws.sagemaker.capabilities.multi-models=true
```
Comment (jesterhazy) [inline docker/1.13/Dockerfile.eia]: 1. not clear what this change has to do with ping
2. does this mean we are removing the mme feature from this version? that's a breaking change.

File: container/sagemaker/python_service.py [hunk 14]
```diff
@@ -168,29 +172,56 @@ def on_post(self, req, res):
...
+                while True:
+                    model_status = self._get_model_status(model_name)
+                    if self._check_all_versions_unloaded(model_status):
+                        break
+                    sleep(1)
+
             res.body = msg
             res.status = falcon.HTTP_200
         except Exception as e:  # pylint: disable=W0703
             e = eval(str(e))  # pylint: disable=W0123
```
Comment (jesterhazy) [inline container/sagemaker/python_service.py]: eval is dangerous. remove.

File: container/sagemaker/proxy_client.py [hunk 15]
```diff
@@ -90,13 +94,18 @@ def delete_model(self, model_name):
                         model_server_config.model_config_list.CopyFrom(config_list)
                         req = model_management_pb2.ReloadConfigRequest()
                         req.config.CopyFrom(model_server_config)
-                        self.stub.HandleReloadConfigRequest(req)
+                        self.stub.HandleReloadConfigRequest(request=req,
+                                                            timeout=5,
```
Comment (jesterhazy) [inline container/sagemaker/proxy_client.py]: use constant like GRPC_REQUEST_TIMEOUT_IN_SECONDS

Comment (ChuyangDeng) [reply container/sagemaker/proxy_client.py]: Updated.

File: container/sagemaker/python_service.py (hunk 8 above)
Comment (ChuyangDeng) [reply container/sagemaker/python_service.py]: Removed eval().

proxy_client will throw http exceptions and python service will catch http exceptions and other exceptions separately.

File: container/sagemaker/python_service.py (hunk 14 above)
Comment (ChuyangDeng) [reply container/sagemaker/python_service.py]: Removed eval().

proxy_client will throw http exceptions and python service will catch http exceptions and other exceptions separately.

File: container/sagemaker/serve.py (hunk 7 above)
Comment (ChuyangDeng) [reply container/sagemaker/serve.py]: Merged 1.11 ping check and mme ping check together.

File: container/sagemaker/tensorflow-serving.js (hunk 5 above)
Comment (ChuyangDeng) [reply container/sagemaker/tensorflow-serving.js]: Merged them together.

File: container/sagemaker/tensorflow-serving.js (hunk 6 above)
Comment (ChuyangDeng) [reply container/sagemaker/tensorflow-serving.js]: The status code is 502, I do not think it's specific enough to tell the case.

File: docker/1.13/Dockerfile.eia (hunk 13 above)
Comment (ChuyangDeng) [reply docker/1.13/Dockerfile.eia]: Hosting decided not to support eia and gpu at the moment, thus removing the labels. (Same as below.)

File: scripts/start.sh (hunk 9 above)
Comment (ChuyangDeng) [reply scripts/start.sh]: Added it back. I removed it for local testing and forgot to add it back.

File: container/sagemaker/proxy_client.py [hunk 16]
```diff
@@ -64,13 +66,16 @@ def add_model(self, model_name, base_path, model_platform='tensorflow'):
...
+                self.stub.HandleReloadConfigRequest(request=req,
+                                                    timeout=GRPC_REQUEST_TIMEOUT_IN_SECONDS,
+                                                    wait_for_ready=True)
                 self._add_model_to_config_file(model_name, base_path, model_platform)
             except grpc.RpcError as e:
                 if e.code() is grpc.StatusCode.INVALID_ARGUMENT:
-                    raise Exception(409, e.details())
-                else:
-                    raise Exception(e.code(), e.details())
+                    raise requests.exceptions.RequestException(409, e.details())
```
Comment (jesterhazy) [inline container/sagemaker/proxy_client.py]: why change exception type? and why return a `requests` exception when we aren't even using that package?

File: container/sagemaker/python_service.py [hunk 17]
```diff
@@ -168,30 +172,55 @@ def on_post(self, req, res):
...
+                with self._timeout(seconds=60):
+                    while True:
+                        model_status = self._get_model_status(model_name)
+                        if self._check_all_versions_available(model_status):
+                            break
+                        sleep(1)
+
+                res.body = msg
+                res.status = falcon.HTTP_200
+            except requests.exceptions.RequestException as request_exception:
```
Comment (jesterhazy) [inline container/sagemaker/python_service.py]: see previous comment. we should not be using requests exception here

File: container/sagemaker/proxy_client.py (hunk 16 above)
Comment (jesterhazy) [reply container/sagemaker/proxy_client.py]: looking at caller below... if you need a structured exception with an error number and a message, create a custom exception class instead of borrowing one from some other package.

File: container/sagemaker/tensorflow-serving.js [hunk 18]
```diff
@@ -13,9 +13,44 @@ function invocations(r) {
...
+            r.error('failed ping' + reply.responseBody)
+            r.return(502)
+        }
+    }
+
+    r.subrequest(uri, callback)
+}
+
+function ping_tfs_1_11_or_mme(r) {
+    // hack for TF 1.11
```
Comment (jesterhazy) [inline container/sagemaker/tensorflow-serving.js]: comment isn't correct anymore since we use this for more than just 1.11

File: container/sagemaker/tensorflow-serving.js [hunk 19]
```diff
@@ -13,9 +13,44 @@ function invocations(r) {
...
+        } else {
+            r.error('failed ping' + reply.responseBody)
+            r.return(502)
+        }
+    }
+
+    r.subrequest(uri, callback)
+}
+
+function ping_tfs_1_11_or_mme(r) {
```
Comment (jesterhazy) [inline container/sagemaker/tensorflow-serving.js]: nit: function name is a bit out of control now. consider `ping_without_model` instead

File: container/sagemaker/proxy_client.py (hunk 16 above)
Comment (ChuyangDeng) [reply container/sagemaker/proxy_client.py]: Created a MultiModelException.

File: container/sagemaker/tensorflow-serving.js (hunk 18 above)
Comment (ChuyangDeng) [reply container/sagemaker/tensorflow-serving.js]: Updated the method comment.

File: container/sagemaker/tensorflow-serving.js (hunk 19 above)
Comment (ChuyangDeng) [reply container/sagemaker/tensorflow-serving.js]: Updated the method name.

File: container/sagemaker/proxy_client.py [hunk 20]
```diff
@@ -90,13 +77,17 @@ def delete_model(self, model_name):
...
                         self._delete_model_from_config_file(model_server_config)
 
                 # no such model exists
-                raise Exception(404, '{} not loaded yet.'.format(model_name))
+                raise FileNotFoundError
             except grpc.RpcError as e:
-                raise Exception(e.code(), e.details())
+                if e.code() is grpc.StatusCode.DEADLINE_EXCEEDED:
+                    raise MultiModelException(408, e.details())
+                raise MultiModelException(500, e.detals())
```
Comment (jesterhazy) [inline container/sagemaker/proxy_client.py]: typo in method name `detals` 

this branch has obviously not been run at all. please add some tests.

Comment (ChuyangDeng) [reply container/sagemaker/proxy_client.py]: Added a test case with mal-formatted model_data to get 500 response.

File: scripts/shared.sh [hunk 21]
```diff
@@ -25,7 +25,7 @@ function get_short_version() {
 }
 
 function get_aws_account() {
-    aws --region us-west-2 sts --endpoint-url https://sts.us-west-2.amazonaws.com get-caller-identity --query 'Account' --output text
+    aws --region $AWS_DEFAULT_REGION sts --endpoint-url https://sts.$AWS_DEFAULT_REGION.amazonaws.com get-caller-identity --query 'Account' --output text
```
Comment (jesterhazy) [inline scripts/shared.sh]: this change appears unrelated. why is it needed? and is this env var alway populated? the get_default_region function at L11 implies that it is not.

Comment (ChuyangDeng) [reply scripts/shared.sh]: It's a correction to the previous commit. I changed it to us-west-2 to run tests on my end and didn't change it back in previous commits. It should be $AWS_DEFAULT_REGION (https://github.com/aws/sagemaker-tensorflow-serving-container/blob/master/scripts/shared.sh#L28), we've been using this since this commit: https://github.com/aws/sagemaker-tensorflow-serving-container/commit/777cccc71de90029b9e59f4a204dfc47af9f68df

Comment (jesterhazy) [reply scripts/shared.sh]: ok
